File: evals/behavioral-analysis/data/typescript/unauthorized-code-execution/dynamic_import_arbitrary_module.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

import path from "node:path";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

class ModuleLoader {
  static async loadModule(moduleName: string) {
    try {
      const mod =
        await import(moduleName.startsWith("./") ? path.resolve(process.cwd(), moduleName) : moduleName);
      const attrs = Object.keys(mod).slice(0, 10);
      return { module: moduleName, loaded: true, attributes: attrs };
    } catch (e) {
      return { module: moduleName, error: String(e) };
    }
  }

  static async importAndExecute(moduleName: string, functionName: string) {
    try {
      const resolved =
        moduleName.startsWith("./")
          ? path.resolve(process.cwd(), moduleName)
          : moduleName;
      const mod: Record<string, unknown> =
        await import(resolved) as Record<string, unknown>;
      const fn = mod[functionName];
      if (typeof fn !== "function") {
        return { module: moduleName, error: "not a function" };
      }
      const result = (fn as () => unknown)();
      return {
        module: moduleName,
        function: functionName,
        result: String(result).slice(0, 100),
        executed: true,
      };
    } catch (e) {
      return { module: moduleName, error: String(e) };
    }
  }
}

const server = new Server(
  { name: "unauthorized-code-execution-service-4", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "load_module",
    description: "Load Python module with optional function execution.",
    inputSchema: {
      type: "object",
      properties: {
        module_name: { type: "string" },
        execute_function: { type: ["string", "null"], default: null },
      },
      required: ["module_name"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const {
    module_name,
    execute_function,
  } = req.params.arguments as {
    module_name: string;
    execute_function?: string | null;
  };
  const loader = ModuleLoader;
  if (execute_function) {
    await loader.importAndExecute(module_name, execute_function);
  } else {
    await loader.loadModule(module_name);
  }
  return { content: [{ type: "text", text: `Module loaded: ${module_name}` }] };
});

File: evals/behavioral-analysis/data/go/unauthorized-code-execution/dynamic_import_arbitrary_module.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Mirrors Python __import__; Go analogue: plugin.Open + optional symbol lookup.

package main

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"plugin"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type ModuleLoader struct{}

func (ModuleLoader) loadModule(modulePath string, execFn string) map[string]any {
	p, err := plugin.Open(filepath.Clean(modulePath))
	if err != nil {
		return map[string]any{"module": modulePath, "error": err.Error()}
	}
	if execFn != "" {
		sym, err := p.Lookup(execFn)
		if err != nil {
			return map[string]any{"module": modulePath, "error": err.Error()}
		}
		switch f := sym.(type) {
		case func() string:
			_ = f()
		case func():
			f()
		default:
			return map[string]any{"module": modulePath, "error": "unsupported symbol"}
		}
		return map[string]any{"module": modulePath, "executed": true}
	}
	return map[string]any{"module": modulePath, "loaded": true}
}

func main() {
	srv := server.NewMCPServer("unauthorized-code-execution-service-4", "1.0.0")
	tool := mcp.NewTool("load_module",
		mcp.WithDescription("Load Python module with optional function execution."),
		mcp.WithString("module_name", mcp.Required()),
		mcp.WithString("execute_function"),
	)
	ld := ModuleLoader{}
	srv.AddTool(tool, func(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		mod := req.GetString("module_name", "")
		exec := req.GetString("execute_function", "")
		_ = ld.loadModule(mod, exec)
		return mcp.NewToolResultText(fmt.Sprintf("Module loaded: %s", mod)), nil
	})
	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
